File: python_automation/src/scrapers/basic_scraper.py
"""
Example of a basic web scraper using requests and BeautifulSoup.
"""
import requests
from bs4 import BeautifulSoup
import logging
import urllib.parse
from typing import Dict, List, Union, Optional, Any
import time
import random


class BasicScraper:
    """Base class for web scraping with common methods."""
    
    def __init__(self, base_url: str, headers: Optional[Dict[str, str]] = None):
        """
        Initialize the scraper.
        
        Args:
            base_url: Base URL for the scraper
            headers: Custom HTTP headers
        """
        self.base_url = base_url
        self.headers = headers or {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        self.session = requests.Session()
        self._setup_logging()
        
        # Optional rate limiting
        self.rate_limit = False
        self.min_delay = 1  # Minimum delay between requests (seconds)
        self.max_delay = 3  # Maximum delay between requests (seconds)
        self.last_request_time = 0
    
    def _setup_logging(self) -> None:
        """Set up the logging system."""
        self.logger = logging.getLogger(self.__class__.__name__)
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)
            self.logger.setLevel(logging.INFO)
    
    def enable_rate_limiting(self, min_delay: float = 1.0, max_delay: float = 3.0) -> None:
        """
        Enable rate limiting for requests to be more respectful to servers.
        
        Args:
            min_delay: Minimum delay between requests in seconds
            max_delay: Maximum delay between requests in seconds
        """
        self.rate_limit = True
        self.min_delay = min_delay
        self.max_delay = max_delay
        self.logger.info(f"Rate limiting enabled: {min_delay}-{max_delay}s between requests")
    
    def _apply_rate_limiting(self) -> None:
        """Apply rate limiting if enabled."""
        if not self.rate_limit:
            return
            
        now = time.time()
        elapsed = now - self.last_request_time
        
        if elapsed < self.min_delay:
            delay = self.min_delay + random.uniform(0, self.max_delay - self.min_delay)
            self.logger.debug(f"Rate limiting: sleeping for {delay:.2f}s")
            time.sleep(delay)
        
        self.last_request_time = time.time()
    
    def get_page(self, url: str) -> Optional[BeautifulSoup]:
        """
        Get and parse the content of a page.
        
        Args:
            url: URL of the page to fetch
            
        Returns:
            BeautifulSoup object or None if error occurs
        """
        try:
            # Validate URL
            parsed = urllib.parse.urlparse(url)
            if not parsed.scheme or not parsed.netloc:
                self.logger.error(f"Invalid URL: {url}")
                return None
                
            # Apply rate limiting if enabled
            self._apply_rate_limiting()
            
            # Make the request
            self.logger.info(f"Fetching page: {url}")
            response = self.session.get(
                url, 
                headers=self.headers, 
                timeout=30,
                allow_redirects=True
            )
            response.raise_for_status()
            
            # Parse with BeautifulSoup
            soup = BeautifulSoup(response.text, 'html.parser')
            self.logger.info(f"Successfully fetched page: {url}")
            return soup
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Error fetching page {url}: {str(e)}")
            return None
        except Exception as e:
            self.logger.error(f"Unexpected error fetching {url}: {str(e)}")
            return None
    
    def resolve_url(self, base_url: str, relative_url: Optional[str]) -> Optional[str]:
        """
        Resolve relative URLs to absolute URLs.
        
        Args:
            base_url: Base URL for resolution
            relative_url: Relative URL to resolve
            
        Returns:
            Absolute URL or None for invalid inputs
        """
        if not relative_url:
            return None
            
        try:
            return urllib.parse.urljoin(base_url, relative_url)
        except Exception as e:
            self.logger.warning(f"Error resolving URL {relative_url} with base {base_url}: {str(e)}")
            return None
    
    def extract_data(self, soup: BeautifulSoup) -> List[Dict[str, Any]]:
        """
        Extract data from a BeautifulSoup object.
        
        Args:
            soup: BeautifulSoup object of the page
            
        Returns:
            List of dictionaries with extracted data
        """
        # Must be implemented by subclasses
        raise NotImplementedError("The extract_data method must be implemented by subclasses")


class NewsScraper(BasicScraper):
    """Specialized scraper for news sites."""
    
    def extract_data(self, soup: BeautifulSoup) -> List[Dict[str, Any]]:
        """
        Extract titles and links of news articles.
        
        Args:
            soup: BeautifulSoup object of the page
            
        Returns:
            List of dictionaries with titles and links
        """
        articles = []
        for article in soup.select('article'):
            title_elem = article.select_one('h2') or article.select_one('h3')
            link_elem = article.select_one('a')
            
            if title_elem and link_elem:
                title = title_elem.get_text(strip=True)
                link = link_elem.get('href')
                
                # Resolve relative URLs
                if link:
                    link = self.resolve_url(self.base_url, link)
                
                # Get description if available
                description_elem = article.select_one('p') or article.select_one('.description')
                description = description_elem.get_text(strip=True) if description_elem else None
                
                # Get date if available
                date_elem = article.select_one('.date') or article.select_one('time')
                date = date_elem.get_text(strip=True) if date_elem else None
                
                # Get image if available
                img_elem = article.select_one('img')
                img_url = None
                if img_elem and img_elem.get('src'):
                    img_url = self.resolve_url(self.base_url, img_elem.get('src'))
                
                if title and link:
                    articles.append({
                        'title': title,
                        'link': link,
                        'description': description,
                        'date': date,
                        'image_url': img_url
                    })
        
        self.logger.info(f"Extracted {len(articles)} articles")
        return articles


def example_usage():
    """Example usage of the news scraper."""
    # Replace with a real news site URL
    scraper = NewsScraper('https://example-news-site.com')
    
    # Enable rate limiting to be respectful
    scraper.enable_rate_limiting(2.0, 5.0)
    
    # Fetch and parse the page
    soup = scraper.get_page('https://example-news-site.com/latest')
    
    if soup:
        # Extract articles
        articles = scraper.extract_data(soup)
        
        # Display the first 5 articles
        for i, article in enumerate(articles[:5], 1):
            print(f"{i}. {article['title']} - {article['link']}")
            if article['description']:
                print(f"   {article['description'][:100]}...")
            print()


if __name__ == "__main__":
    # Configure logging
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )
    
    # Run the example
    example_usage() 

File: nodejs_automation/src/scrapers/basicScraper.js
/**
 * Basic web scraper implementation using Axios and Cheerio.
 */
const axios = require('axios');
const cheerio = require('cheerio');
const logger = require('../utils/logger');
const fileUtils = require('../utils/fileUtils');
const path = require('path');
const url = require('url');  // Node.js URL module for safer URL handling

/**
 * Base class for scrapers.
 */
class BaseScraper {
  /**
   * Initializes the scraper.
   * @param {string} baseUrl - Base URL for the scraper
   * @param {Object} options - Additional options
   */
  constructor(baseUrl, options = {}) {
    this.baseUrl = baseUrl;
    this.options = options;
    
    // Default axios configuration
    this.axiosConfig = {
      headers: {
        'User-Agent': options.userAgent || 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml',
        'Accept-Language': 'en-US,en;q=0.9',
        ...options.headers
      },
      timeout: options.timeout || 30000,
      maxRedirects: options.maxRedirects || 5,
      validateStatus: status => status >= 200 && status < 300, // Only accept 2xx status codes
    };

    // Add optional rate limiting
    this.requestDelay = options.requestDelay || 0; // Delay in ms between requests
    this.lastRequestTime = 0; // Timestamp of the last request
  }
  
  /**
   * Gets the content of a web page.
   * @param {string} targetUrl - URL of the page
   * @returns {Promise<{$: CheerioStatic, url: string}|null>} Cheerio object and URL or null on error
   */
  async fetchPage(targetUrl) {
    const operation = logger.startOperation(`Fetch page: ${targetUrl}`);
    
    try {
      // Validate URL
      const parsedUrl = new URL(targetUrl);
      if (!['http:', 'https:'].includes(parsedUrl.protocol)) {
        throw new Error('Only HTTP and HTTPS protocols are supported');
      }

      // Apply rate limiting if needed
      const now = Date.now();
      const elapsed = now - this.lastRequestTime;
      if (elapsed < this.requestDelay) {
        const delay = this.requestDelay - elapsed;
        await new Promise(resolve => setTimeout(resolve, delay));
      }
      
      // Send the request
      this.lastRequestTime = Date.now();
      const response = await axios.get(targetUrl, this.axiosConfig);
      const $ = cheerio.load(response.data);
      
      operation.end('success');
      return { $, url: response.config.url };
    } catch (error) {
      operation.error(error);
      
      // Detailed error information
      if (error.response) {
        logger.error(`HTTP Error ${error.response.status}: ${error.response.statusText}`);
      } else if (error.request) {
        logger.error('Server did not respond');
      } else {
        logger.error(`Request error: ${error.message}`);
      }
      
      return null;
    }
  }
  
  /**
   * Resolves relative URLs to absolute URLs.
   * @param {string} baseUrl - Base URL
   * @param {string} relativeUrl - Relative URL
   * @returns {string|null} Absolute URL or null if input is invalid
   */
  resolveUrl(baseUrl, relativeUrl) {
    if (!relativeUrl && relativeUrl !== '') return null;
    
    try {
      // Use the URL API for safer URL resolution
      return new URL(relativeUrl, baseUrl).href;
    } catch (error) {
      logger.warn(`Invalid URL resolution: ${baseUrl} + ${relativeUrl}`);
      return null;
    }
  }
  
  /**
   * Extract data from page. Must be implemented by subclasses.
   * @param {CheerioStatic} $ - Cheerio object loaded with HTML
   * @param {string} url - Page URL
   * @returns {Promise<Array|Object>} Extracted data
   */
  async extractData($, url) {
    throw new Error('The extractData method must be implemented by subclasses');
  }
  
  /**
   * Saves the extracted data to a file.
   * @param {Array|Object} data - Extracted data
   * @param {string} outputPath - Output file path
   * @returns {Promise<boolean>} True if data was saved successfully
   */
  async saveData(data, outputPath) {
    const ext = path.extname(outputPath).toLowerCase();
    
    switch (ext) {
      case '.json':
        return await fileUtils.saveJson(data, outputPath);
      case '.csv':
        return await fileUtils.saveCsv(data, outputPath);
      default:
        logger.error(`Unsupported file format: ${ext}`);
        return false;
    }
  }
  
  /**
   * Runs the scraper on a URL and saves the results.
   * @param {string} url - URL to process
   * @param {string} outputPath - Path to save the results
   * @returns {Promise<{success: boolean, data: Array|Object|null}>} Operation result
   */
  async run(url, outputPath) {
    const operation = logger.startOperation(`Scraper on ${url}`);
    
    try {
      const page = await this.fetchPage(url);
      if (!page) {
        operation.end('failed to get page');
        return { success: false, data: null };
      }
      
      const data = await this.extractData(page.$, page.url);
      
      if (outputPath) {
        await this.saveData(data, outputPath);
      }
      
      operation.end(`success (${Array.isArray(data) ? data.length : 1} items)`);
      return { success: true, data };
    } catch (error) {
      operation.error(error);
      return { success: false, data: null };
    }
  }
}

/**
 * Specialized scraper for news sites.
 */
class NewsScraper extends BaseScraper {
  /**
   * Extracts news articles from a page.
   * @param {CheerioStatic} $ - Cheerio object
   * @param {string} url - Page URL
   * @returns {Promise<Array>} Array of news articles
   */
  async extractData($, url) {
    const articles = [];
    
    // Generic selector for news blocks - adjust as needed
    $('article, .news-item, .article, .post').each((i, el) => {
      // Look for title
      const titleElement = $(el).find('h1, h2, h3, .title').first();
      const title = titleElement.text().trim();
      
      // Look for link
      const linkElement = titleElement.find('a').length > 0 
        ? titleElement.find('a') 
        : $(el).find('a').first();
      
      const link = this.resolveUrl(url, linkElement.attr('href'));
      
      // Look for description
      const description = $(el).find('p, .description, .summary').first().text().trim();
      
      // Look for date
      const dateText = $(el).find('.date, time, .published').first().text().trim();
      
      // Look for image
      const imageUrl = this.resolveUrl(
        url, 
        $(el).find('img').first().attr('src')
      );
      
      // Only add if title and link are present
      if (title && link) {
        articles.push({
          title,
          link,
          description: description || null,
          date: dateText || null,
          imageUrl: imageUrl || null
        });
      }
    });
    
    logger.info(`Extracted ${articles.length} articles from ${url}`);
    return articles;
  }
}

/**
 * Example usage of the scraper.
 */
async function example() {
  // Create a scraper instance
  const scraper = new NewsScraper('https://example-news.com');
  
  // Define output path
  const outputPath = path.join('data', `news-${fileUtils.getTimestamp()}.json`);
  
  // Run the scraper
  const result = await scraper.run('https://example-news.com/latest', outputPath);
  
  if (result.success) {
    logger.info(`Scraping completed successfully: ${outputPath}`);
    
    // Show the first 3 results
    const sample = result.data.slice(0, 3);
    logger.info(`Data sample: ${JSON.stringify(sample, null, 2)}`);
  } else {
    logger.error('Scraping failed');
  }
}

module.exports = {
  BaseScraper,
  NewsScraper,
  example
}; 